### PR TITLE
Update erc721 docs

### LINKED ISSUE #563: update erc721 docs

### PR DESCRIPTION
**Pull request by andrew-fleming**
Supersedes #699 and fixes #563.

### CODE REVIEW

File: src/token/erc721/erc721.cairo [hunk 1]
```diff
@@ -125,6 +144,8 @@ mod ERC721 {
 
     #[external(v0)]
     impl ERC721MetadataCamelOnlyImpl of interface::IERC721MetadataCamelOnly<ContractState> {
+        /// Camel case support.
+        /// See [token_uri](token_uri).
```
Comment (ericnordelo) [inline src/token/erc721/erc721.cairo:148]:
```suggestion
    /// Adds camelCase support for `IERC721Metadata`.
    #[external(v0)]
    impl ERC721MetadataCamelOnlyImpl of interface::IERC721MetadataCamelOnly<ContractState> {
```

File: src/token/erc721/erc721.cairo [hunk 2]
```diff
@@ -133,26 +154,32 @@ mod ERC721 {
...
 
+        /// Query if `operator` is an authorized operator for `owner`.
         fn is_approved_for_all(
             self: @ContractState, owner: ContractAddress, operator: ContractAddress
         ) -> bool {
             self.ERC721_operator_approvals.read((owner, operator))
         }
 
+        /// Change or reaffirm the approved address for an NFT.
+        /// Emits an [Approval](Approval) event.
```
Comment (ericnordelo) [inline src/token/erc721/erc721.cairo:182]:
```suggestion
        /// Emits an `Approval` event.
```

This syntax of `[]()` is for solidity doc gen tools, and we shouldn't replicate it (or we should document it).

Applies for the rest of the file as well.

File: src/token/erc721/erc721.cairo [hunk 3]
```diff
@@ -164,12 +191,17 @@ mod ERC721 {
             self._approve(to, token_id);
         }
 
+        /// Enable or disable approval for `operator` to manage all of the
+        /// caller's assets.
+        /// Emits an [Approval](Approval) event.
```
Comment (ericnordelo) [inline src/token/erc721/erc721.cairo:196]:
```suggestion
        /// Enable or disable approval for `operator` to manage all of the
        /// caller's assets.
        ///
        /// Emits an [Approval](Approval) event.
```

Solidity's library is doing this, and I think it reads better. I would separate the short description, from the long one, and from the requirements, and events, etc... with a line.

File: src/token/erc721/erc721.cairo [hunk 4]
```diff
@@ -195,34 +231,48 @@ mod ERC721 {
 
     #[external(v0)]
     impl ERC721CamelOnlyImpl of interface::IERC721CamelOnly<ContractState> {
+        /// Camel case support.
+        /// See [balance_of](balance_of).
```
Comment (ericnordelo) [inline src/token/erc721/erc721.cairo:235]: This is why I think it makes sense to add just one comment on top of the impl block, to avoid repeating these comments that don't add much.

For consistency we should add the

`/// Adds camelCase support for...`

on the impl, instead of each method.

File: docs/modules/ROOT/pages/api/erc721.adoc [hunk 5]
```diff
@@ -0,0 +1,654 @@
...
+
+See <<IERC721Metadata-token_uri,IERC721Metadata::token_uri>>.
+
+[.contract-item]
+[[ERC721-balanceOf]]
+==== `[.contract-item-name]#++balanceOf++#++(self: @ContractState, account: ContractAddress) -> u256++` [.item-kind]#external#
+
+See <<IERC721-balance_of,IERC721::balance_of>>.
+
+Supports the Cairo v0 convention of writing external methods in camelCase as discussed {casing-discussion}.
```
Comment (ericnordelo) [inline docs/modules/ROOT/pages/api/erc721.adoc:403]: I don't think there's much value in repeating this on every camelCase method. I vote for removing them.

**Review by ericnordelo — COMMENTED** [review]
Looking good! Left some comments.

File: docs/modules/ROOT/pages/erc721.adoc [hunk 6]
```diff
@@ -1,772 +1,361 @@
...
-== IERC721
-
-[,cairo]
+:token-types: https://docs.openzeppelin.com/contracts/4.x/tokens#different-kinds-of-tokens[non-fungible tokens]
+:eip721: https://eips.ethereum.org/EIPS/eip-721[EIP-721]
+:erc721-api: xref:/api/erc721.adoc[API Reference]
+:introspection: xref:/introspection.adoc[Introspection]
+
+The ERC721 token standard is a specification for {token-types}, or more colloquially: NFTs.
+The `token::erc721::ERC721` contract implements an approximation of {eip721} in Cairo for Starknet.
```
Comment (martriay) [inline docs/modules/ROOT/pages/erc721.adoc:9]:
```suggestion
The ERC721 token standard is a specification for {token-types}, or more colloquially: NFTs.
The `token::erc721::ERC721Component` component implements an approximation of {eip721} in Cairo for Starknet.
```

File: docs/modules/ROOT/pages/erc721.adoc [hunk 7]
```diff
@@ -1,772 +1,361 @@
...
+The following preset contract includes all three interfaces and represents the full interface of the `ERC721ABI` module.
+
+WARNING: The {ierc721-interface} and {ierc721metadata-interface} interfaces are approximations of their Ethereum counterparts.
+These approximations include a few notable differences.
+See {compatibility} for more information.
+
+[,javascript]
+----
+trait IERC721ABI {
+    // IERC721
```
Comment (martriay) [inline docs/modules/ROOT/pages/erc721.adoc:31]: now that we have components and presets, I wonder if we should begin with this chunk of interface. It's kindof big and with lots of information (non standard, camel support). maybe i'd focus in the component. what do you think?

File: docs/modules/ROOT/pages/erc721.adoc [hunk 8]
```diff
@@ -1,772 +1,361 @@
...
-@contract_interface
-namespace IERC721Receiver {
-    func onERC721Received(
-        operator: felt, from_: felt, tokenId: Uint256, data_len: felt data: felt*) -> (selector: felt) {
+:components: https://community.starknet.io/t/cairo-1-contract-syntax-is-evolving/94794#extensibility-and-components-11[Components]
+:mint-api: xref:api/erc721.adoc#ERC721-_mint[_mint]
+
+WARNING: The following example uses a contract's `unsafe_new_contract_state` to access another contract's state.
+Although this is useful to use them as modules, it's considered unsafe because storage members could clash among used contracts if not reviewed carefully.
+Extensibility will be revisited after {components} are introduced.
```
Comment (martriay) [inline docs/modules/ROOT/pages/erc721.adoc:177]: all this section needs to be updated to components

File: docs/modules/ROOT/pages/erc721.adoc [hunk 9]
```diff
@@ -1,772 +1,361 @@
...
+Implementing the `IERC721Receiver` interface exposes the {on_erc721_received-api} method.
+When safe methods such as {safe_transfer_from-api} and {safe_mint-api} are called, they invoke the recipient contract's `on_erc721_received` method which *must* return the {receiver-id}.
+Otherwise, the transaction will fail.
 
-Returns a token ID at a given `index` of all the tokens stored by the contract.
-Use along with <<totalsupply,totalSupply>> to enumerate all tokens.
+TIP: For information on how to calculate interface IDs, see {computing-interface-id}.
 
-Parameters:
+==== Creating a token receiver contract
```
Comment (martriay) [inline docs/modules/ROOT/pages/erc721.adoc:305]: this also needs to be updated to components

File: docs/modules/ROOT/pages/api/erc721.adoc [hunk 10]
```diff
@@ -0,0 +1,648 @@
...
+* xref:#IERC721-owner_of[`++owner_of(token_id)++`]
+* xref:#IERC721-safe_transfer_from[`++safe_transfer_from(from, to, token_id, data)++`]
+* xref:#IERC721-transfer_from[`++transfer_from(from, to, token_id)++`]
+* xref:#IERC721-approve[`++approve(to, token_id)++`]
+* xref:#IERC721-set_approval_for_all[`++set_approval_for_all(operator, approved)++`]
+* xref:#IERC721-get_approved[`++get_approved(token_id)++`]
+* xref:#IERC721-is_approved_for_all[`++is_approved_for_all(owner, operator)++`]
+
+.ISRC5
+* xref:/api/introspection.adoc#ISRC5-supports_interface[`++supports_interface(interface_id)++`]
```
Comment (martriay) [inline docs/modules/ROOT/pages/api/erc721.adoc:44]: this is not actually part of the `IERC721` trait

File: docs/modules/ROOT/pages/api/erc721.adoc [hunk 11]
```diff
@@ -0,0 +1,648 @@
...
+
+Requirements:
+
+- `operator` cannot be the caller.
+
+[.contract-item]
+[[ERC721-_mint]]
+==== `[.contract-item-name]#++_mint++#++(ref self: ContractState, to: ContractAddress, token_id: u256)++` [.item-kind]#internal#
+
+WARNING: Usage of this method is discouraged, use <<IERC721-_safe_mint,_safe_mint>> whenever possible.
```
Comment (martriay) [inline docs/modules/ROOT/pages/api/erc721.adoc:519]: same as in the ERC721 PR, why is this discouraged? they're different with their tradeoffs, imo none is inherently more secure than the other

**Review by martriay — COMMENTED** [review]
this needs to be updated to use components, but other than that it looks good!

File: docs/modules/ROOT/pages/api/erc721.adoc (hunk 11 above)
Comment (andrew-fleming) [reply docs/modules/ROOT/pages/api/erc721.adoc:519]: Valid point, will update

File: docs/modules/ROOT/pages/erc721.adoc (hunk 7 above)
Comment (andrew-fleming) [reply docs/modules/ROOT/pages/erc721.adoc:31]: I agree that it is kind of big with a lot going on. I'm not sure what you're suggesting though. Do you mean removing the `Interface` section entirely? Or removing the individual interface subsections and just keep the `ERC721ABI`?

File: docs/modules/ROOT/pages/erc721.adoc [hunk 12]
```diff
@@ -1,772 +1,277 @@
...
+
+The ERC721 token standard is a specification for {token-types}, or more colloquially: NFTs.
+`token::erc721::ERC721Component` provides an approximation of {eip721} in Cairo for Starknet.
+
+== Interface
+
+:compatibility: xref:/erc721.adoc#erc721_compatibility[ERC721 Compatibility]
+:ierc721-interface: xref:/erc721.adoc#ierc721[IERC721]
+:ierc721metadata-interface: xref:/erc721.adoc#ierc721metadata[IERC721Metadata]
+:dual-interfaces: xref:interfaces.adoc##dual_interfaces[Dual interfaces]
```
Comment (ericnordelo) [inline docs/modules/ROOT/pages/erc721.adoc:14]:
```suggestion
:dual-interfaces: xref:interfaces.adoc#dual_interfaces[Dual interfaces]
```

File: docs/modules/ROOT/pages/erc721.adoc [hunk 13]
```diff
@@ -1,772 +1,277 @@
...
+[,javascript]
+----
+#[starknet::contract]
+mod MyNFT {
+    use openzeppelin::introspection::src5::SRC5Component;
+    use openzeppelin::token::erc721::ERC721Component;
+    use starknet::ContractAddress;
+
+    component!(path: SRC5Component, storage: src5, event: SRC5Event);
+    component!(path: ERC721Component, storage: erc721, event: ERC721Event);
```
Comment (ericnordelo) [inline docs/modules/ROOT/pages/erc721.adoc:107]: I think we should swap the order (this applies to the impls below as well). This is compliant with the account preset too.

File: docs/modules/ROOT/pages/erc721.adoc [hunk 14]
```diff
@@ -1,772 +1,277 @@
...
 ----
+#[starknet::contract]
+mod MyTokenReceiver {
+    use openzeppelin::introspection::src5::SRC5Component;
+    use openzeppelin::token::erc721::ERC721ReceiverComponent;
+    use starknet::ContractAddress;
 
-==== `tokenOfOwnerByIndex`
+    component!(path: SRC5Component, storage: src5, event: SRC5Event);
+    component!(path: ERC721ReceiverComponent, storage: erc721_receiver, event: ERC721ReceiverEvent);
```
Comment (ericnordelo) [inline docs/modules/ROOT/pages/erc721.adoc:232]: Same comment regarding swapping

File: docs/modules/ROOT/pages/api/erc721.adoc [hunk 15]
```diff
@@ -0,0 +1,599 @@
...
+[.contract-index]
+.Events
+--
+.IERC721
+* xref:#IERC721-Approval[`++Approval(owner, approved, token_id)++`]
+* xref:#IERC721-ApprovalForAll[`++ApprovalForAll(owner, operator, approved)++`]
+* xref:#IERC721-Transfer[`++Transfer(from, to, token_id)++`]
+--
+
+==== External functions
```
Comment (ericnordelo) [inline docs/modules/ROOT/pages/api/erc721.adoc:255]:
```suggestion
==== Embeddable functions
```

**Review by ericnordelo — COMMENTED** [review]
Looking good Andrew! Left a couple of small comments but I think it looks ready.

**Review by ericnordelo — APPROVED** [review]
LGTM!

File: src/token/erc721/interface.cairo [hunk 16]
```diff
@@ -137,57 +137,3 @@ trait IERC721ReceiverCamel<TState> {
         data: Span<felt252>
     ) -> felt252;
 }
-
-//
-// ERC721 ABI
```
Comment (martriay) [inline src/token/erc721/interface.cairo:62]: why was this removed? is it related to merging https://github.com/OpenZeppelin/cairo-contracts/pull/761/files?

File: docs/modules/ROOT/pages/erc721.adoc [hunk 17]
```diff
@@ -2,28 +2,20 @@
 
 :token-types: https://docs.openzeppelin.com/contracts/4.x/tokens#different-kinds-of-tokens[non-fungible tokens]
```
Comment (martriay) [inline docs/modules/ROOT/pages/erc721.adoc:3]:
```suggestion
:token-types: https://docs.openzeppelin.com/contracts/5.x/tokens#different-kinds-of-tokens[non-fungible tokens]
```
bump solidity docs link

File: docs/modules/ROOT/pages/erc721.adoc (hunk 7 above)
Comment (martriay) [reply docs/modules/ROOT/pages/erc721.adoc:23]: i'm not sure i understood the second option, did you mean `IERC721`? i was suggesting to remove the interface section, although it may require a bit more of thought. My impression is that it's not that important now we have components because it's not a single module with all of it anymore, now you can select the impls.

If anything, it is closer to the preset, which is no yet documented and probably not the main focus of the section either.

File: docs/modules/ROOT/pages/api/erc721.adoc [hunk 18]
```diff
@@ -627,6 +576,8 @@ See <<IERC721-Transfer,Transfer>>.
 use openzeppelin::token::erc721::interface::IERC721Receiver;
 ```
 
+Interface for any contract that wants to support safe token transfers from ERC721 asset contracts.
```
Comment (martriay) [inline docs/modules/ROOT/pages/api/erc721.adoc:579]:
```suggestion
Interface for contracts that support receiving `safe_transfer_from` transfers.
```

File: docs/modules/ROOT/pages/erc721.adoc [hunk 19]
```diff
@@ -304,48 +216,52 @@ TIP: For information on how to calculate interface IDs, see {computing-interface
 
 ==== Creating a token receiver contract
 
+The Contracts for Cairo `IERC721ReceiverImpl` already returns the correct interface ID for safe token transfers.
+To integrate the `IERC721Receiver` interface into a contract, simply include the ABI embed directive to the implementation and add the `initializer` in the contract's constructor .
```
Comment (martriay) [inline docs/modules/ROOT/pages/erc721.adoc:220]: crazy idea, would it make sense to offer the `IERC721Receiver` impl as part of the `ERC721Component` instead of a separate one?

**Review by martriay — APPROVED** [review]
Looks good! Left minor comments.

I'm having troubles compiling the snippet from the usage section. Most probably an error on my end, but couldn't yet pinpoint it. Can you confirm it works?



I'm using scarb 2.3.1 and I'm using `main`

File: src/token/erc721/interface.cairo (hunk 16 above)
Comment (andrew-fleming) [reply src/token/erc721/interface.cairo]: Yes, exactly

File: docs/modules/ROOT/pages/erc721.adoc (hunk 7 above)
Comment (andrew-fleming) [reply docs/modules/ROOT/pages/erc721.adoc:23]: > i'm not sure i understood the second option, did you mean IERC721?

Yeah, that section, `IERC721Metadata`, and `ISRC5`. I removed those section already since I left that response

> i was suggesting to remove the interface section, although it may require a bit more of thought. My impression is that it's not that important now we have components because it's not a single module with all of it anymore, now you can select the impls.

> If anything, it is closer to the preset, which is no yet documented and probably not the main focus of the section either. The focus should be the component and design/implementation details.

Agreed^

Maybe it'd make sense to follow a structure like this in the doc:

`ERC721Component`
- `Implementations`
  - `ERC721Impl`
  - `ERC721MetadataImpl`
  ...
- `ERC721 compatibility`
- `Usage`
...

We can either head the `impl`s with the interface snippet or just provide a link to the interface in the API...or just remove the `Interface`/`Implementations` section entirely, idk

**Comment by andrew-fleming** [comment]
> I'm having troubles compiling the snippet from the usage section. Most probably an error on my end, but couldn't yet pinpoint it. Can you confirm it works?

Weird. Yeah, I copied the .toml you posted and I'm also using Scarb 2.3.1 and it compiles on my end. I don't have the syntax highlighting though

File: docs/modules/ROOT/pages/erc721.adoc (hunk 7 above)
Comment (martriay) [reply docs/modules/ROOT/pages/erc721.adoc:23]: Right now we have this inconsistent structures:

**Upgrades**
1. explainer
2. usage

**Account**
1. interface

**Access**
1. usage
2. interface

**ERC20**
1. interface
4. compat
5. usage

**ERC721**
2. interface
3. compat
4. usage

**Security**
1. usage

**Introspection**
1. interface
2. usage

Comment (martriay) [reply docs/modules/ROOT/pages/erc721.adoc:23]: maybe we should have

1. explainer
2. usage
3. compatibility

and leave interface for API, of have it between explainer and usage

File: docs/modules/ROOT/pages/erc721.adoc (hunk 19 above)
Comment (martriay) [reply docs/modules/ROOT/pages/erc721.adoc:220]: bad idea

**Review by martriay — APPROVED** [review]
It works now, I had to delete my `Scarb.lock`.
